crates/router/src/compatibility/wrap.rs

```diff
@@ -40,7 +40,28 @@ where
         api_authentication,
     )
     .await;
-    match resp {
+    handle_application_response::<Q, S, E>(request, resp)
+}
+
+pub fn handle_application_response<Q, S, E>(
+    request: &HttpRequest,
+    server_resp: Result<api::ApplicationResponse<Q>, error_stack::Report<E>>,
+) -> HttpResponse
+where
+    Q: Serialize + std::fmt::Debug,
+    E: Serialize + error_stack::Context + actix_web::ResponseError + Clone,
+    S: TryFrom<Q> + Serialize,
+    error_stack::Report<E>: services::EmbedError,
+{
+    match server_resp {
+        Ok(api::ApplicationResponse::ResponseWithCustomHeader(response, headers)) => {
+            let mut final_response = handle_application_response::<Q, S, E>(request, Ok(*response));
```

Comment:
Why the function calling itself?

Reply:
`ResponseWithCustomHeader` has two fields. The first field is `response` of type `ApplicationResponse`, and the second field is `headers` which is a `HeaderMap`. The function `handle_application_response` takes an `ApplicationResponse` as input and returns an `HttpResponse`. To obtain the `HttpResponse` for the inner `response`, the function is recursively called again. Similar behavior can be observed in other functions as well.